### PR TITLE
Add more command line arguments

### PR DESCRIPTION
**Pull request by anish-mudaraddi**
simple change to add command line arguments to specify where to save output files and log files etc

```
muspinsim --help

usage: __main__.py [-h] [-o OUT_DIR] [-l LOG_PATH] [-f FITREPORT_PATH]
                   input_file

Muspinsim - A program designed to carry out spin dynamics calculations for
muon science experiments

positional arguments:
  input_file            filepath to muspinsim specially formatted text file
                        specifying simulation input parameters.

options:
  -h, --help            show this help message and exit
  -o OUT_DIR, --out-dir OUT_DIR
                        folder to store the output .dat files
  -l LOG_PATH, --log-path LOG_PATH
                        filepath to store simulation logs
  -f FITREPORT_PATH, --fitreport-path FITREPORT_PATH
                        filepath to store fit report if fitting parameters
                        given
```

### CODE REVIEW

File: muspinsim/__main__.py
Comment (elichad) [inline muspinsim/__main__.py:139]: rep_fname needs some default too – running without this arg raises `UnboundLocalError: local variable 'rep_fname' referenced before assignment`

File: muspinsim/__main__.py
Comment (elichad) [inline muspinsim/__main__.py:89]: We shouldn't assume that a user will always enter a path that includes directories. Using `-l test.log` raises `NotADirectoryError`

File: muspinsim/__main__.py
Comment (elichad) [inline muspinsim/__main__.py:138]: Same issue as the log file here - if the user doesn't include a directory this raises `NotADirectoryError`

File: muspinsim/__main__.py
Comment (elichad) [inline muspinsim/__main__.py:27]: This function name could be clearer that it will create directories if they don't exist. Functions with names like `check_x` often act only as validation and tend not to manipulate things, so this is a slightly confusing break from (admittedly not super well-established) convention. I'd suggest something like `ensure_dir_path_exists` instead.

File: muspinsim/__main__.py
Comment (elichad) [inline muspinsim/__main__.py:50]: Another common convention is to have optional arguments in alphabetical order (aside from `-h` which is sometimes kept at the top). Makes it easier if you're looking for something specific in the help section.

**Review by elichad — CHANGES_REQUESTED** [review]

**Comment by elichad** [comment]
This could also do with some tests – maybe more integration-type tests where you model a user putting in certain command-line arguments, and then make sure outputs are written to the correct locations etc.

(sorry for the delayed review, btw)

File: test/test_args.py
Comment (anish-mudaraddi) [inline test/test_args.py:64]: There seems to be a bug where the log file doesn't get created when this test is run.

File: test/test_args.py
Comment (elichad) [inline test/test_args.py:48]: This is fine as-is, but are you familiar with [f-strings](https://realpython.com/python-f-strings/)? It's a matter of personal preference but I like them for readability. Suggestion is an example of what they'd look like in this case, you don't have to commit it.
```suggestion
                    f"{out_tmp_dir}", 
                    "-l",
                    f"{out_tmp_dir}/new_test.log",
```

File: test/test_args.py
Comment (elichad) [reply test/test_args.py:64]: I see commented lines like this in every test - do they all have this problem?

File: test/test_args.py
Comment (elichad) [inline test/test_args.py:20]: Good work adding these tests :)

**Review by elichad — CHANGES_REQUESTED** [review]

File: muspinsim/__main__.py
Comment (anish-mudaraddi) [reply muspinsim/__main__.py:99]: yep that's a mistake is should just be
`log_path = os.path.join(inp_dir, args.log_path)`

File: muspinsim/__main__.py
Comment (anish-mudaraddi) [reply muspinsim/__main__.py:104]: yea that makes sense - we can just change this to
`os.path.join(os.path.dirname(log_path), os.path.basename(log_path))`

File: muspinsim/__main__.py
Comment (anish-mudaraddi) [reply muspinsim/__main__.py]: i just removed this comment

**Comment by anish-mudaraddi** [comment]
I fixed the issue with log files not being created

previously when running:
```
logging.basicConfig()
```

 if it found any existing log handlers, it would return without doing anything
 
 I added the `force` flag to ignore log handlers that `unittest` adds.

File: muspinsim/__main__.py
Comment (patrick-austin) [inline muspinsim/__main__.py:159]:
```suggestion
                try:            
                    # check if directory exists, if not create it
                    ensure_dir_path_exists(os.path.dirname(args.fit_report_path))
                    rep_fname = os.path.basename(args.fit_report_path)
                    rep_dname = os.path.dirname(args.fit_report_path)
                except NotADirectoryError:
                    # if fit_report_path was just a filename use inp_dir as a default
                    rep_fname = args.fit_report_path
                    rep_dname = inp_dir
```
If above change to comments accepted, make this consistent.

File: muspinsim/__main__.py
Comment (patrick-austin) [reply muspinsim/__main__.py:75]: I find the suggested change easier to read/understand.

**Review by patrick-austin — COMMENTED** [review]
Only have minor comments on readability of code (comments). @joelvdavies I'd value a second opinion on my suggested changes - if you think it's better as it currently is I'm happy to approve this now, or if you think something else would be more readable we can go with that. It's my first time looking at mu spin sim code so don't want my unfamiliarity with the codebase to result in pointless nitpicking.

**Pull request merged**
